lib/documents.js

```diff
@@ -2716,6 +2717,284 @@ function urisOutputTransform(headers, data) {
+    throw new Error('consistentSnapshot needs to be a boolean.');
+  }
+  requester.startRequest(operation).result(mlutil.callbackOn(jobState, onReadAllInit))
+      .catch((err) => {
+        jobState.inputStream.emit('error', err);
+        jobState.requesterCount = 0;
+        jobState.error = err;
+        finishReader(jobState);
+      });
+
+  return  jobState.duplexStream;
+}
+
+function onReadAllInit(output) {
+  let maxRequesters = 0;
+  const jobState = this;
+  switch (jobState.jobOptions.concurrentRequests.multipleOf) {
+    case 'hosts':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier *
+          new Set(output.map((forest) => forest.host)).size;
+      break;
+    default:
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier * output.length;
+      break;
+  }
+  jobState.startTime = Date.now();
+  for (let i = 0; i < maxRequesters; i++) {
+    jobState.requesterCount++;
+    onReadAllDocs(jobState,i);
+  }
+}
+
+function onReadAllDocs(jobState, readerId) {
+
+  if(jobState.error){
+    finishReader(jobState);
+    return;
+  }
+  let readBatchArray = [];
+  const batchSize = jobState.jobOptions.batchSize;
+  jobState.lastBatch = false;
+  jobState.retryCount = 100;
+
+  if(jobState.jobOptions.writableObjectMode === 'array'){
+    readBatchArray = jobState.inputStream.read();
+  }
+  else {
+    a:
+        for (let i = 0; i < batchSize; i++) {
+          const uri = jobState.inputStream.read();
+          if (uri === null) {
+            switch (readBatchArray.length) {
+              case 0: {
+                finishReader(jobState);
+                break a;
+              }
+              default: {
+                jobState.lastBatch = true;
+                break a;
+              }
+            }
+          }
+          readBatchArray.push(uri);
+        }
+  }
+  if (readBatchArray.length !== 0) {
+    const batchdef = {uris: readBatchArray};
+    if(jobState.jobOptions.transform){
+      batchdef.transform = jobState.jobOptions.transform;
+    }
+    if(jobState.jobOptions.consistentSnapshot){
+      if(!jobState.consistentSnapshotTimestamp){
+        jobState.consistentSnapshotTimestamp = new Date();
+      }
+      batchdef.timestamp = jobState.consistentSnapshotTimestamp;
+    }
+    if(jobState.jobOptions.categories){
+      batchdef.categories = jobState.jobOptions.categories;
+    }
+    readDocs(jobState, false, batchdef,readBatchArray, readerId);
+  }
+}
+
+function readDocs(jobState, val, batchdef, readBatchArray, readerId){
+  if(jobState.error){
+    finishReader(jobState);
+    return;
+  }
+  readDocumentsImpl.call(jobState.docInstance, val, [batchdef])
+      .result((output) => {
+        jobState.docsReadSuccessfully+= readBatchArray.length;
+        output.forEach(item => {
+          if(jobState.readableObjectMode === 'chunked'){
+            jobState.outputStream.write(JSON.stringify(item));
```

Comment:
As discussed in email, it would be much better in the chunking case to 

- read the response to the request as a stream
- listen for data events on the response stream and write data events to the job output stream
- listen for error events on the response stream with error disposition logic
- listen for the end event to determine when to fire the next request

Avoiding stringify is important for performance reasons, which is at a premium for bulk data movement.